enterprise/internal/database/perms_store.go

```diff
@@ -2042,3 +2046,149 @@ func sliceToSet[T comparable](s []T) map[T]struct{} {
+	if err != nil {
+		return nil, err
+	}
+
+	if authzParams.BypassAuthz {
+		// NOTE(naman): If Authz conds were bypassed then no need to fetch
+		// perm details for the repos. Show permission reason as "Site Admin"
+		// or "Authentication Bypassed" based on query params used.
+
+		reason := UserRepoPermissionReasonAuthzBypass
+
+		if authzParams.BypassAuthzReason == database.BypassAuthzReasonSiteAdmin {
+			reason = UserRepoPermissionReasonSiteAdmin
+		}
+
+		for _, repo := range repos {
+			perms = append(perms, &UserPermission{Repo: repo, Reason: reason})
+		}
+	} else {
+		// NOTE(naman): If Authz conds were not bypassed then find perm details
+		// for each of the repos and show correct reason and updatedAt.
+
+		repoIDs := []int32{}
+		for _, repo := range repos {
+			repoIDs = append(repoIDs, int32(repo.ID))
+		}
+
+		permissionsInfoMap, err := s.getRepoPermissionInfoForIDs(ctx, userID, repoIDs)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, repo := range repos {
+			var updatedAt time.Time
+			reason := UserRepoPermissionReasonUnrestricted
+
+			info := permissionsInfoMap[int32(repo.ID)]
+			if info != nil {
+				reason = info.Reason
+				updatedAt = info.UpdatedAt
+			}
+
+			perms = append(perms, &UserPermission{Repo: repo, Reason: reason, UpdatedAt: updatedAt})
+		}
+	}
+
+	return perms, err
+}
+
+type UserPermission struct {
+	Repo      *types.Repo
+	Reason    UserRepoPermissionReason
+	UpdatedAt time.Time
+}
+
+type UserRepoPermissionReason string
+
+// UserRepoPermissionReason constants.
+const (
+	UserRepoPermissionReasonAuthzBypass    UserRepoPermissionReason = "Authentication Bypassed"
+	UserRepoPermissionReasonSiteAdmin      UserRepoPermissionReason = "Site Admin"
+	UserRepoPermissionReasonUnrestricted   UserRepoPermissionReason = "Unrestricted"
+	UserRepoPermissionReasonPermissionSync UserRepoPermissionReason = "Permission Sync"
+)
+
+type ListUserPermissionsArgs struct {
+	Query          string
+	PaginationArgs *database.PaginationArgs
+}
+
+func (s *permsStore) getRepoPermissionInfoForIDs(ctx context.Context, userID int32, repoIDs []int32) (map[int32]*repoPermissionInfo, error) {
+	permissionInfoQuery := sqlf.Sprintf(permissionInfoQueryFmt, userID, pq.Array(repoIDs))
+	rows, err := s.Query(ctx, permissionInfoQuery)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { err = basestore.CloseRows(rows, err) }()
+
+	infoMap := make(map[int32]*repoPermissionInfo, len(repoIDs))
+	for rows.Next() {
+		info, err := scanRepoPermInfo(rows)
+		if err != nil {
+			return infoMap, err
+		}
+
+		infoMap[info.RepoID] = info
+	}
+
+	return infoMap, nil
+}
+
+const permissionInfoQueryFmt = `
+SELECT
+	id,
+	up.updated_at AS permission_updated_at,
+	CASE 
+		WHEN up.user_id IS NOT NULL THEN 'Permission Sync'
+		ELSE 'Unrestricted'
+	END AS permission_reason
+FROM
+	repo
+	LEFT JOIN user_permissions AS up ON up.user_id = %d
+		AND up.object_ids_ints @> INTSET (repo.id)
+WHERE
+	repo.id = ANY (%s)
+`
```

Comment:
The query is not looking at the external service, which might also be marked as `unrestricted` in case permissions are not enabled. 

I think this query would give us what we want:
```suggestion
-- restricted repos
SELECT unnest(object_ids_ints) AS id, 'Permission Sync' AS permission_reason
  FROM user_permissions AS up
  WHERE up.user_id  = %d
UNION
-- unrestricted repos
SELECT repo_id AS id, 'Unrestricted' AS permission_reason
  FROM repo_permissions
  WHERE unrestricted
`

And we need to do a different query if `globals.PermissionsUserMapping().Enabled` is `false`, like we do in `authzQuery`.

And we can do a completely different query for site admins if their permissions are not enforced (e.g. just paginate through repos and give `site_admin` as a reason).
```

Reply:
All of that is handled by the `Repos().List()`.